File: rust/src/bin/p0076.rs
// Project Euler: Problem 76

euler::run_solver!(76);

fn solve() -> String {
    compute((1..100).collect(), 100).to_string()
}

fn compute(coins: Vec<usize>, target: usize) -> usize {
    let mut tbl = vec![0_usize; target + 1];
    tbl[0] = 1;

    for c in coins {
        for i in c..=target {
            tbl[i] += tbl[i - c];
        }
    }
    tbl[target]
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0076_5() {
        assert_eq!(compute((1..5).collect(), 5), 6);
    }
    #[test]
    fn p0076_100() {
        assert_eq!(compute((1..100).collect(), 100), 190569291);
    }
}

File: ocaml/src/p0076.ml
(* Project Euler: Problem 76 *)

(*
  another version of the problem 31

  coins: 1, 2, 3, ..., 99
  total: 100
*)

open Core

let compute coins target =
  let table = Array.create ~len:(target + 1) 0 in
  let rec aux = function
    | [] -> table.(target)
    | x :: xs ->
      for i = x to target do
        table.(i) <- table.(i) + table.(i - x)
      done;
      aux xs
  in
  table.(0) <- 1;
  aux coins
;;

let solve () = compute (List.range 1 99 ~stop:`inclusive) 100 |> Int.to_string
;;

(* Test *)

let%test_unit "5" =
  [%test_eq: int] (compute (List.range 1 4 ~stop:`inclusive) 5) 6

let%test_unit "100" =
  [%test_eq: int] (compute (List.range 1 99 ~stop:`inclusive) 100) 190569291
;;
